src/source/channel_router.rs

```diff
@@ -0,0 +1,306 @@
+// Channel router types and implementation.
+
+use crate::{Sample, Source};
+use std::{
+    cmp::min,
+    error::Error,
+    fmt,
+    sync::mpsc::{channel, Receiver, Sender},
+};
+
+/// A matrix to map inputs to outputs according to a gain
+///
+/// A two-dimensional matrix of `f32`s:
+/// - The first dimension is respective to the input channels
+/// - The second is respective to the output channels
+///
+/// Thus, if a value at `map[1,1]` is 0.2, this signifies that the signal on
+/// channel 1 should be mixed into channel 1 with a coefficient of 0.2.
+pub type ChannelMap = Vec<Vec<f32>>;
+// doing this as Vec<Vec<atomic_float::AtomicF32>> would require feature=experimental, so I decided
+// to just use a channel to do updates.
+//
+// Doing it as a HashMap<(u16,u16), f32> is an option too but there's a penalty hashing these
+// values, there's ways to speed that up though. It'd be great if the object upgraded its
+// implementation if it got sufficiently big.
+
+// pub fn empty_channel_map(inputs: u16, outputs: u16) -> ChannelMap {
+//     vec![vec![0.0f32; outputs.into()]; inputs.into()]
+// }
+
+/// Internal function that builds a [`ChannelRouter<I>`] object.
+pub fn channel_router<I>(
+    input: I,
+    channel_count: u16,
+    channel_map: ChannelMap,
+) -> (ChannelRouterController, ChannelRouterSource<I>)
+where
+    I: Source,
+    I::Item: Sample,
+{
+    ChannelRouterSource::new(input, channel_count, channel_map)
+}
+
+struct ChannelRouterMessage(usize, usize, f32);
+
+/// `ChannelRouterController::map()` returns this error if the router source has been dropped.
+#[derive(Debug, Eq, PartialEq)]
+pub struct ChannelRouterControllerError {}
+
+impl fmt::Display for ChannelRouterControllerError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "<ChannelRouterControllerError>")
+    }
+}
+
+impl Error for ChannelRouterControllerError {}
+
+/// A controller type that sends gain updates to a corresponding [`ChannelRouterSource`].
+#[derive(Debug, Clone)]
+pub struct ChannelRouterController {
+    sender: Sender<ChannelRouterMessage>,
+}
+
+impl ChannelRouterController {
+    /// Set or update the gain setting for a channel mapping.
+    ///
+    /// A channel from the input may be routed to any number of channels in the output, and a
+    /// channel in the output may be a mix of any number of channels in the input.
+    ///
+    /// Successive calls to `map` with the same `from` and `to` arguments will replace the
+    /// previous gain value with the new one.
+    pub fn map(
+        &mut self,
+        from: u16,
+        to: u16,
+        gain: f32,
+    ) -> Result<(), ChannelRouterControllerError> {
```

Comment:
:question:  If the routing has to change one will have to remember to clear all the gains that are not needed anymore. 
Maybe a "reset" function that sets all the gains to zero would be helpful.

:question: I wonder if we should allow specifying matrix at the initialization time, or let users to pass a list of coefficients.

Reply:
Probably that is why I wanted to set the whole gain set at once: users may have the matrix on their side process and inspect it a they prefer and then use its to update the router.